---
title: "chr6analysis"
author: "Whitney Domico"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


cleaning data from NCBI Chr 6 Homology
```{r}
Chr6DataNCBI <- read_excel("C:/Users/Whitn/OneDrive/Desktop/BIO247/BIO247Project/BIO247ProjectArticles/Chr6DataNCBI.xlsx")
Chr6DataNCBI <- Chr6DataNCBI[,1:3]

temp <- c()
generang <- c()
for (each in Chr6DataNCBI$`human position`){
  temp <- unlist(strsplit(each, ':'))
  temp2 <- temp[2]
  temp3 <- unlist(strsplit(temp2, " "))
  temp4 <- temp3[1]
  generang <- c(generang, temp4)
}

Chr6DataNCBI$gene.range <- generang

startpos <- c()
endpos <- c()
for (each in Chr6DataNCBI$gene.range){
  temp <- unlist(strsplit(each, "-"))
  temp2 <- gsub(",", "", temp)
  startpos <- c(startpos, as.numeric(temp2[1])/1000000)
  endpos <- c(endpos, as.numeric(temp2[2])/1000000)
}

Chr6DataNCBI$startpos <- startpos
Chr6DataNCBI$endpos <- endpos
```


finding most common SNP range
```{r}
row <- c(1:length(MainData$SNP))
chr6 <- c()
for (each in row){
  if (MainData[each,2]=="06"){
    chr6 <- c(chr6, as.numeric(MainData[each,3]))
  }
}

library(ggplot2)
testdf <- data.frame("initial"= chr6)
testdf$binnedchr6 <- cut(chr6, c(0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170))
ggplot(testdf)+geom_bar(aes(x=binnedchr6))


row <- c(1:length(MainData$SNP))
chr6 <- c()
SNP6 <- c()
for (each in row){
  if (MainData[each,2]=="06"){
    if (MainData[each,3]>20 && MainData[each,3]<=40){
      chr6 <- c(chr6, as.numeric(MainData[each,3]))
      SNP6 <- c(SNP6, MainData[each, 'SNP'])
    }
  }
}
data.frame("SNP"=SNP6,"Position.bp"=chr6)

freqtable6 <- as.data.frame(table(testdf$binnedchr6))
mean <- mean(freqtable6$Freq)
sd <- sd(freqtable6$Freq)

onesd <- 1+sd
twosd <- 1+2*sd

over68 <- c()
over95 <- c()

row <- c(1:length(freqtable6$Freq))

for (each in row){
  if (freqtable6[each,'Freq']> onesd && freqtable6[each,'Freq'] < twosd){
      over68 <- c(over68, freqtable6[each,'Var1'])
  } else if (freqtable6[each,'Freq'] >= twosd){
      over95 <- c(over95, freqtable6[each,'Var1'] )
  }
}
print(over68)
print(over95)

```

so 3rd and 4th columns are significant

finding genes in Chr6 that include the bp with SNP
```{r}
potSNP6 <- c()
potgenes6 <- c()
row <- c(1:length(Chr6DataNCBI$startpos))
for (each in chr6){
  for (each2 in row){
    if (each >= Chr6DataNCBI[each2,'startpos'] && each <= Chr6DataNCBI[each2,'endpos']){
        potSNP6 <- c(potSNP6, each)
        potgenes6 <- c(potgenes6, Chr6DataNCBI[each2,3])
    }
  }
}
unqpotgenes6<-unique(potgenes6)

Chr6Genes <- data.frame(potSNP6)
Chr6Genes$potgenes6 <- potgenes6
```