---
title: "BIO247 Lab 6"
author: "Whitney Domico"
date: "`r Sys.Date()`"
output:
  word_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
https://www.ncbi.nlm.nih.gov/projects/homology/maps/human/chr6/
```

```{r data, include=FALSE}
Conc_data <- read.csv("C:/Users/Whitn/Downloads/Conc_data.csv")
```

**Figure 1.**

The figure below shows the change in Protein A concentration over a 100ms time period. There appears to be little correlation, and therefore it could be suggested that Protein A has a constant concentration.

```{r plot1, echo=FALSE}
plot(Conc_data$Time..ms., Conc_data$X.ProtA., xlab= "Time (ms)", ylab=("[Protein A]"), pch=19, col='black')
```

**Figure 2.**

The figure below shows the change in concentration of four proteins (Protein A, Protein B, Protein C, Protein D) over a 100ms time period. The proteins are represented by scatterplots in colors black, orange, green, and purple, respectively. Proteins A and C appear to have little average change over the time period, while Proteins B and C do appear to have sort of positive correlation with time. Due to Protein B's higher concentration as compared to the others, the scale makes this plot difficult to read. See Figure 4 for separate correlational graphs for the proteins.

```{r plot2, echo=FALSE}
plot(Conc_data$Time..ms., Conc_data$X.ProtA., xlab= "Time (ms)", ylab=("Concentration"), pch=16, col='black', ylim=c(0,40.5))
points(Conc_data$Time..ms., Conc_data$X.ProtB., pch=16, col='orange')
points(Conc_data$Time..ms., Conc_data$X.ProtC., pch=16, col='green')
points(Conc_data$Time..ms., Conc_data$X.ProtD., pch=16, col='purple')
```

**Figure 3.**

This plot, shown below, is similar to Figure 3 in that it also shows concentration of Proteins A, B, C, and D over a 100ms time period. Again, there appears to be no correlation between concentration and time for Proteins A and C, and there appears to be positive correlations between concentration and time for Proteins B and D. The line graph element of this figure makes these assertions more visible.

```{r plot3, echo=FALSE}
plot(Conc_data$Time..ms., Conc_data$X.ProtA., xlab= "Time (ms)", ylab=("Concentration"), pch=19, col='black', ylim=c(0,40.5), type='l', lwd=2.5)
lines(Conc_data$Time..ms., Conc_data$X.ProtB., pch=19, col='orange', lwd=2.5)
lines(Conc_data$Time..ms., Conc_data$X.ProtC., pch=19, col='green', lwd=2.5)
lines(Conc_data$Time..ms., Conc_data$X.ProtD., pch=19, col='purple', lwd=2.5)
legend('topleft', c("Protein A", "Protein B", "Protein C", "Protein D"), lty=1,lwd=2.5, col=c('black', 'orange', 'green', 'purple'))
```

**Figure 4.**

The figure below includes the same data as Figures 2 and 3; however, it splits the proteins onto separate graphs of their own concentrations versus the 100ms time period. In this depiction, it is more obvious that Proteins A and C appear to have no correlation between concentration and time and may therefore be treated as constant over time. Protein B is again a positive correlation, but this depiction shows it more visibly as a positive and perhaps linear relationship. Similarly, Protein D again shows a positive correlation, but this depiction shows it more visibly as a positive and perhaps logarithmic relationship approaching a horizontal asymptote of 1.0.

```{r plot4, echo=FALSE}
old.par <- par(mfrow=c(2,2), mar=c(3,4.5,3,1))
plot(Conc_data$Time..ms., Conc_data$X.ProtA., xlab= "Time (ms)", ylab=("[Protein A]"), type='l', lty=1, col='black')
plot(Conc_data$Time..ms., Conc_data$X.ProtB., xlab= "Time (ms)", ylab=("[Protein B]"), type='l',lty=1, col='orange')
plot(Conc_data$Time..ms., Conc_data$X.ProtC., xlab= "Time (ms)", ylab=("[Protein C]"), type='l',lty=1, col='green')
plot(Conc_data$Time..ms., Conc_data$X.ProtD., xlab= "Time (ms)", ylab=("[Protein D]"),type='l', lty=1, col='purple')
par(old.par)
```

**Figure 5.**

This figure, shown below, again depicts concentration of Protein B over time. A linear regression line has been added in black. The line fits well and supports the assumption that Protein B increases linearly with time; however, it cannot be ignored that perhaps the concentration of Protein B has a different shape past the time constraint of 100ms. Therefore, more experimentation with wider time frames must be executed.

```{r plot5, echo=FALSE}
plot(Conc_data$Time..ms., Conc_data$X.ProtB., xlab= "Time (ms)", ylab=("[Protein B]"), pch=16, col='orange')
abline(lm(Conc_data$X.ProtB. ~ Conc_data$Time..ms.), lwd=2)
```

**Figure 6.**

The figure below again shows concentration of Protein D over time. A non-linear regression line of the Michaelis-Menten equation ((a\*time)/(b+Time)) was added. This line does not visually appear to fit well; however, it returns a correlational coefficient of 0.97, which suggests a good fit. The data appears to approach a horizontal asymptote of 1.0, which may be expected to be the maximum possible concentration of Protein D.

```{r plot6, echo=FALSE}
fit <- nls(Conc_data$X.ProtD.~a*Conc_data$Time..ms./(b+Conc_data$Time..ms.), data=Conc_data, start=list(a=0.1,b=0.1))
plot(Conc_data$Time..ms., Conc_data$X.ProtD., xlab= "Time (ms)", ylab=("[Protein D]"), pch=19, col='purple')
lines(Conc_data$Time..ms., predict(fit), lwd=2)
```

```{r PMCs, include=FALSE}
library(readxl)
PMC3077530_Data <- read.csv("C:/Users/Whitn/OneDrive/Desktop/BIO247/BIO247Project/BIO247ProjectArticles/PMC3077530_Data.txt", sep="", header = TRUE)
PMC2775422_Data <- read.csv("C:/Users/Whitn/OneDrive/Desktop/BIO247/BIO247Project/BIO247ProjectArticles/PMC2775422_Data.txt", sep="", header = TRUE)
PMC3912837_Data <- read_excel("C:/Users/Whitn/OneDrive/Desktop/BIO247/BIO247Project/BIO247ProjectArticles/PMC3912837_Data.xlsx",sheet = "Sheet2")
PMC2890845_Data <- read_excel("C:/Users/Whitn/OneDrive/Desktop/BIO247/BIO247Project/BIO247ProjectArticles/PMC2890845_Data.xlsx")
PMC4724864_Data <- read_excel("C:/Users/Whitn/OneDrive/Desktop/BIO247/BIO247Project/BIO247ProjectArticles/PMC4724864_Data.xlsx", skip = 1)
PMC6927206_Data <- read_excel("C:/Users/Whitn/OneDrive/Desktop/BIO247/BIO247Project/BIO247ProjectArticles/PMC6927206_Data.xlsx", skip = 1)
PMC3872086_Data <- read_excel("C:/Users/Whitn/OneDrive/Desktop/BIO247/BIO247Project/BIO247ProjectArticles/PMC3872086_Data.xlsx", skip = 5)
PMC4059435_Data <- read_excel("C:/Users/Whitn/OneDrive/Desktop/BIO247/BIO247Project/BIO247ProjectArticles/PMC4059435_Data.xlsx")
PMC3905728_Data <- read_excel("C:/Users/Whitn/OneDrive/Desktop/BIO247/BIO247Project/BIO247ProjectArticles/PMC3905728_Data.xlsx", sheet = "Sheet1")
PMC3827979_Data <- read_excel("C:/Users/Whitn/OneDrive/Desktop/BIO247/BIO247Project/BIO247ProjectArticles/PMC3827979_Data.xlsx")
library(dplyr)
library(data.table)
require(data.table)
```

**Figure 7.**

The figure below combines data of common SNPs found in patients with schizophrenia and plots the relative frequency of these common SNPs occurring on each chromosome (relative frequency was calculated as ((raw number of unique SNPs/length of chromosome in base pairs)). If the distribution of SNPs was truly random, it would be expected that all chromosomes would have a relative frequency of about 1.0; however, this graph shows that chromosomes 6, 22, and 11 have unusually high relative frequencies, and this has been confirmed through statistical analysis. Therefore, these chromosomes have been selected for further analysis.

```{r plot 7, echo=FALSE}
temp <- substr(PMC3077530_Data$SNP,1,nchar(PMC3077530_Data$SNP)-3)
temp3 <- c()
for (each in PMC3912837_Data$SNP){
  temp2 <- unlist(strsplit(each, ""))
  if ("*" %in% temp2){
   temp3 <- c(temp3, (substr(each, 2, nchar(each))))
  } else
    temp3 <- c(temp3, each)
}
SNP <- c(temp, PMC2775422_Data$SNP, temp3, PMC2890845_Data$`Tag SNP ID`, PMC4724864_Data$SNP, PMC6927206_Data$SNP...1, PMC3872086_Data$SNP, PMC4059435_Data$SNP, PMC3905728_Data$`Strongest SNP`, PMC3827979_Data$`rs ID`)
temp <- substr(PMC3077530_Data$SNP,1,nchar(PMC3077530_Data$SNP)-3)

temp3 <- c()
for (each in PMC3912837_Data$SNP){
  temp2 <- unlist(strsplit(each, ""))
  if ("*" %in% temp2){
   temp3 <- c(temp3, (substr(each, 2, nchar(each))))
  } else
    temp3 <- c(temp3, each)
}
SNP <- c(temp, PMC2775422_Data$SNP, temp3, PMC2890845_Data$`Tag SNP ID`, PMC4724864_Data$SNP, PMC6927206_Data$SNP...1, PMC3872086_Data$SNP, PMC4059435_Data$SNP, PMC3905728_Data$`Strongest SNP`, PMC3827979_Data$`rs ID`)
temp4 <- c()
for (each in PMC2890845_Data$CNVR) {
  temp <- unlist(strsplit(each, ""))
  temp2 <- c(temp[4:5])
  if (":" %in% temp2 && "X" %in% temp2 == FALSE && "Y" %in% temp2 == FALSE && "x" %in% temp2 == FALSE && "y" %in% temp2 == FALSE){
    temp3 <- paste0(0, temp2[1])
    temp4 <- c(temp4, temp3)
  } else if (":" %in% temp2){
    temp4 <- c(temp4, temp2[1])
  } else
    temp5 <- paste0(temp2[1], temp2[2])
    temp4 <- c(temp4, temp5)
temp <- c()
temp2 <- c()
temp3 <- c()
temp5 <- c()
}
PMC2890845_Data$Chr <- temp4

temp4 <- c()
for (each in PMC2890845_Data$CNVR) {
  temp <- unlist(strsplit(each, ":"))
  temp2 <- c(temp[2])
  temp3 <- unlist(strsplit(temp2, "â€“"))
  temp4 <- c(temp4, as.numeric(temp3[1]))
  temp <- c()
  temp2 <- c()
  temp3 <- c()
}
PMC2890845_Data$Position.bp <- temp4

temp4 <- c()
for (each in PMC3827979_Data$ChrPos) {
  temp <- unlist(strsplit(each, ""))
  temp2 <- c(temp[4:5])
  if (":" %in% temp2 && "X" %in% temp2 == FALSE && "Y" %in% temp2 == FALSE && "x" %in% temp2 == FALSE && "y" %in% temp2 == FALSE){
    temp3 <- paste0(0, temp2[1])
    temp4 <- c(temp4, temp3)
  } else if (":" %in% temp2){
    temp4 <- c(temp4, temp2[1])
  } else
    temp5 <- paste0(temp2[1], temp2[2])
    temp4 <- c(temp4, temp5)
temp <- c()
temp2 <- c()
temp3 <- c()
temp5 <- c()
}
PMC3827979_Data$Chr <- temp4
temp <- PMC2775422_Data$Position.bp.
temp <- temp/1000000
temp2 <- PMC3912837_Data$Position.bp/1000000
temp3 <- PMC2890845_Data$Position.bp/1000000
temp5 <- PMC4724864_Data$BP/1000000
temp6 <- PMC6927206_Data$BP...3/1000000
temp7 <- PMC4059435_Data$`Location (bp)`/1000000
temp8 <- as.numeric(PMC3905728_Data$`Position (hg18)`)/1000000
temp9 <- PMC3827979_Data$bp/1000000
Position.Mb <- c(PMC3077530_Data$Position.Mb., temp, temp2, temp3, temp5, temp6, PMC3872086_Data$Mb, temp7, temp8, temp9)
temp4 <- c()
for (each in PMC3912837_Data$Chr) {
  temp <- unlist(strsplit(each, ""))
  temp2 <- c(temp[4:5])
  if (":" %in% temp2 && "X" %in% temp2 == FALSE && "Y" %in% temp2 == FALSE && "x" %in% temp2 == FALSE && "y" %in% temp2 == FALSE){
    temp3 <- paste0(0, temp2[1])
    temp4 <- c(temp4, temp3)
  } else if (":" %in% temp2){
    temp4 <- c(temp4, temp2[1])
  } else
    temp5 <- paste0(temp2[1], temp2[2])
    temp4 <- c(temp4, temp5)
temp <- c()
temp2 <- c()
temp3 <- c()
temp5 <- c()
}
Chr <- c()
temp6 <- c(PMC3077530_Data$Chr, PMC2775422_Data$Chr, temp4, PMC2890845_Data$Chr, PMC4724864_Data$CHR, PMC6927206_Data$CHR, PMC3872086_Data$Chr, PMC4059435_Data$Chr., PMC3905728_Data$Chromosome, PMC3827979_Data$Chr)
for (each in temp6){
  temp7 <- unlist(strsplit(each, ""))
  if (length(temp7) == 1 && "X" %in% temp7 == FALSE && "Y" %in% temp7 == FALSE && "x" %in% temp7 == FALSE && "y" %in% temp7 == FALSE) {
    temp8 <- paste0(0, each)
    Chr <- c(Chr, temp8)
 } else 
    Chr <- c(Chr, each)
temp6 <- c()
}
PMC2890845_Data$PMCID <- c("PMC2890845")
PMC3912837_Data$PMCID <- c("PMC3912837")
PMC2775422_Data$PMCID <- c("PMC2775422")
PMC3077530_Data$PMCID <- c("PMC3077530")
PMC4724864_Data$PMCID <- c("PMC4724864")
PMC6927206_Data$PMCID <- c("PMC6927206")
PMC3872086_Data$PMCID <- c("PMC3872086")
PMC4059435_Data$PMCID <- c("PMC4059435")
PMC3905728_Data$PMCID <- c("PMC3905728")
PMC3827979_Data$PMCID <- c("PMC3827979")
PMCID <- c(PMC3077530_Data$PMCID, PMC2775422_Data$PMCID, PMC3912837_Data$PMCID, PMC2890845_Data$PMCID, PMC4724864_Data$PMCID, PMC6927206_Data$PMCID, PMC3872086_Data$PMCID, PMC4059435_Data$PMCID, PMC3905728_Data$PMCID, PMC3827979_Data$PMCID)
InitialData <- data.frame(SNP, Chr, Position.Mb, PMCID)
library(data.table)
require(data.table)


tempdf2 <- setDT(InitialData)[,.(Chr=paste(Chr,collapse = ",")), by=SNP]

temp2 <- c()
for (each in tempdf2$Chr){
  temp <- unlist(strsplit(each, ","))
  temp2 <- c(temp2, temp[1])
}
 
tempdf3 <- setDT(InitialData)[,.(Position.Mb=paste(Position.Mb,collapse = ",")), by=SNP]

temp3 <- c()
for (each in tempdf3$Position.Mb){
  temp <- unlist(strsplit(each, ","))
  temp3 <- c(temp3, temp[1])
}
 
tempdf4 <- setDT(InitialData)[,.(PMCID=paste(PMCID,collapse = ",")), by=SNP]

MainData <- data.frame("SNP"= tempdf2$SNP, "Chr"=temp2, "Position.Mb"=temp3, "PMCID"=tempdf4$PMCID)
library(readxl)
library(ggplot2)
ChrData <- read_excel("C:/Users/Whitn/OneDrive/Desktop/BIO247/BIO247Project/BIO247ProjectArticles/ChrData.xlsx")


ObsData <- as.data.frame(table(MainData$Chr))
ObsData$ChrBP <- ChrData$`Length (bp)`
ObsData$Chr <- c(ObsData$Var1[1:23], "Y")

ObsData$AdjObs <- 10^7*(ObsData$Freq/ObsData$ChrBP)

library(ggplot2)
ggplot(ObsData, color= station, fill='black', aes(x=Var1, y=AdjObs))+
  geom_col(position='dodge')+
  ylab("Relative Frequency (AU)")+
  xlab("Chromosome")
```
