---
title: "RangeAndSequence"
author: "Casey Hansen"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
if (!require("BiocManager", quietly = TRUE))
install.packages("BiocManager")
BiocManager::install()
BiocManager::install("GenomicRanges")
library(IRanges)
library(ggplot2)
```

## 1 Plotting IRange Data
This chunk plots a set of IRanges data, using more robust methods than in plot-ranges.R

```{r}
# Define Ranges
x <- IRanges(start=c(4, 7, 2, 20), end=c(13, 7, 2, 23))
names(x) <- letters[1:4]

# Housekeeping - defining boundaries so graph looks nice
sep=0.5
height=0.5
# Convert ranges to dataframe for ease of reference
out <- as.data.frame(x)
out$y <- disjointBins(x)
out$ymin <- out$y * (sep + height) - height
out$ymax <- out$ymin + height
xmin <- min(out$start)
xmax <- max(out$end)
xbreaks <- seq(xmin - 1L, xmax + 1L)

# Plot the ranges
ggplot(out) + geom_rect(aes(ymin=ymin, ymax=ymax, xmin=start-0.5, xmax=end+0.5)) + geom_text(aes(x=start + width/2 - 0.5, y=ymin+(ymax-ymin)/2, label=names), size=8, color="grey") + scale_x_continuous(breaks=xbreaks) + xlab("") + ylab("") + theme(axis.text.y=element_blank()) + theme(panel.grid.major=element_blank(), panel.grid.minor.y=element_blank(), axis.ticks=element_blank())
```

## 2 Annontation Data


```{r pressure, echo=FALSE}

# Install necessary packages
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("TxDb.Mmusculus.UCSC.mm10.ensGene")

# Upload Data library, and assign it to a variable
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
txdb <- TxDb.Mmusculus.UCSC.mm10.ensGene

# Retrieve information from uploaded data
promoters(txdb)
length(genes(txdb))
```

## 3 Uploading FASTA data

This first chunk uses the 'readLines()' function, reading each line verbatim; "takes a file and converts every line into an indexable data"; basically a list; not really useful for large data
```{r, echo=FALSE}

# Create placeholder vectors for eventual dataframe
name <- c()
sequence <- c()

# Open file for reading; open="r" means "open for reading"
con <- file('C:/Users/Whitn/OneDrive/Desktop/dubjay-bioinformatics/Chapters9-10/egfr_flank.fasta',open="r")
line <- readLines(con) 

# Go line-by-line, getting the name and sequence, saving in appropriate vector; grep means "starts with"
for (each in line){
  if (grepl(">", each)){
    name <- append(name, each)
    #If you want ONLY the name
    #name <- append(name, strsplit(each,"|",fixed=TRUE)[[1]][1])
  } else if (each == "") {
    next
  } else {
    sequence <- append(sequence, each)
  }
}

# Turn vectors into a dataframe
seq_data = data.frame(name,sequence)

```

The next method uses the 'read.table()' function, filling creating 'NA' entries to fill in empty columns. This method is faster, and good if there is more information in the header line for each sequence, but requires more post-processing.

```{r}
# Read in file, fill in blank spaces so dataframe has correct dimensions
df <- read.table('C:/Users/Whitn/OneDrive/Desktop/dubjay-bioinformatics/Chapters9-10/egfr_flank.fasta',header=FALSE,sep="|",fill=TRUE)
```

The final method is specific for FASTA files, using the 'Biostrings' library and 'readDNAStringSet()'. Most specialized can sometimes mean least robust, though also less prone to error.
```{r}
# Upload library
library("Biostrings")

# Read in FASTA file
fastaFile <- readDNAStringSet('C:/Users/Whitn/OneDrive/Desktop/dubjay-bioinformatics/Chapters9-10/egfr_flank.fasta')

# Access names and sequences to put into a dataframe
seq_name = names(fastaFile)
sequence = paste(fastaFile)
BS_data = data.frame(seq_name,sequence)
```

## 4 FASTQ Data
Here, we go through the laborious process of
1. uploading a FASTQ file using 'read.table'
2. creating placeholder vectors for desired information
3. transforming the sequence scores from their character format to their PHRED scores
```{r,echo=FALSE}

# Upload FASTQ file, identify indices of name, sequence, and score information
fastq <- read.table('C:/Users/Whitn/OneDrive/Desktop/dubjay-bioinformatics/Chapters9-10/contam.fastq',header=FALSE,sep="|",fill=TRUE)
name <- seq(1, length(fastq$V1), by=4)
seq <- seq(2, length(fastq$V1), by=4)
score <- seq(4, length(fastq$V1), by=4)

# Save name, sequence, and score character information into vectors, create placeholder for score value
Name <- fastq$V1[name]
Sequence <- fastq$V1[seq]
Score_char <- fastq$V1[score]
Score_val = c()

#Create initial data frame
fastq_data <- data.frame(Name, Sequence, Score_char)

# Split each score character string
# Transform score from character to ASCII number to PHRED number
# Save sequence scores in list in Score_val
for (x in 1:length(fastq_data$Name)){
  score_num = c()
  char <- unlist(strsplit(fastq_data$Score_char[x],""))
  for (a in char){
    score_num <- append(score_num,utf8ToInt(a)-33)
  }
  Score_val <- append(Score_val,list(score_num))
}

# Add each list of score values to a column in the dataframe
for (i in 1:(length(Score_val))){
  fastq_data$Score_val[i] <- list(Score_val[i])
}

```

